Fix editor area/actor deselection bugs

drogoganor opened the pull request:
There are two state issues with selection and deselection in the editor after the new changes.

* After dragging out an area selection, right clicking to deselect the area leaves the area selection panel visible.
* If you drag out an area selection, then click an actor, then click cancel, the area selection panel is still shown.

This PR fixes these issues. `ActorEditLogic` needed some logic to fire in a handler for `DefaultBrush.SelectionChanged` instead of `Tick()`.

Also added a missing copyright header.

Also fixes #21314

drogoganor committed:
Fix editor area/actor deselection bugs

Porenutak commented:
a bit off topic, but it would be intuitive if selected area can be canceled also with Escape key. Same for Copy/paste brush

PunkPun commented on OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs:
```diff
@@ -194,8 +187,7 @@ public override void Tick()
 					var font = Game.Renderer.Fonts[typeLabel.Font];
-					var truncatedType = WidgetUtils.TruncateText(actor.DescriptiveName, typeLabel.Bounds.Width, font);
-					typeLabel.GetText = () => truncatedType;
+					typeLabel.Text = WidgetUtils.TruncateText(actor.DescriptiveName, typeLabel.Bounds.Width, font);
```
please don't regress #21309

drogoganor replied:
Hey Pun, this change in commit 2fde98a (https://github.com/OpenRA/OpenRA/pull/21300) is the cause of issue #21314 . 
I think it's an issue with measuring the bounds when GetText() is invoked.
I've backed out this change and will address it separately after I've fixed the undo issue.

PunkPun replied:
in that case this was not the correct fix, instead this should be done

```csharp
var truncatedType = WidgetUtils.TruncateText(TranslationProvider.GetString(actor.DescriptiveName), typeLabel.Bounds.Width, font);
typeLabel.GetText = () => truncatedType;
```

PunkPun reviewed (commented):
undo does not seem to remove actor selection visual

drogoganor commented on OpenRA.Mods.Common/EditorBrushes/EditorDefaultBrush.cs:
```diff
@@ -84,6 +84,11 @@ public void ClearSelection()
 
+		public void UpdateSelection()
+		{
+			SelectionChanged?.Invoke();
```
The undo/redo needs to trigger the SelectionChanged event, so I've added this public function to invoke it. I'm not sure if it's the best way of going about it and feedback would be appreciated.

PunkPun replied:
you could change the Selection variable to always invoke selection changed event when set

```csharp
EditorSelection selection = new();
public EditorSelection Selection
{
	get => selection;
	set
	{
		selection = value;
		SelectionChanged?.Invoke();
	}
}
```

drogoganor commented on OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs:
```diff
@@ -356,12 +349,32 @@ ScrollItemWidget DropdownSetup(KeyValuePair<string, string> option, ScrollItemWi
 		{
-			if (CurrentActor != null)
-				editorActionManager.Add(new RemoveActorAction(editorActorLayer, CurrentActor));
+			var selectedActor = CurrentActor;
```
Since selection changes are included in undo/redo actions, we first need to take a reference to the currently selected actor, deselect it, then delete it. Otherwise, when we undo, we try to select an actor which is already deleted.

PunkPun commented:
the history tab seems a bit bugged. If you go to a select action then you are put into the selection panel, instead of being able to move through the rest on history. I dunno what the best way to fix that is.

And also all subsequent actions are deleted as the "Cleared Selection" action is added (or is going back to history tab inserting the action?)

drogoganor commented:
I'll check out that History tab bug. Sorry Pun I thought it was in better shape.
EDIT: That history tab issue is fixed in the most recent commit now.

There's a couple of other code design issues I still want to address too:
* The ActorEditLogic has a CurrentActor property. I think this should actually come from the `EditorDefaultBrush.Selection.Actor` now.
* Additionally, there's a Selected property on the EditorActorPreview. I'm wondering if this flag should be set to true/false in the EditorDefaultBrush now instead.

PunkPun commented:
yeah, probably they should be controlled by the default brush

PunkPun commented on OpenRA.Mods.Common/Widgets/Logic/Editor/MapEditorTabsLogic.cs:
```diff
@@ -63,12 +63,6 @@ void SetupTab(string buttonId, string tabId, MenuType tabType)
-				editor.SetBrush(editor.DefaultBrush);
-				editor.DefaultBrush.ClearSelection();
-			}
-
```
perhaps a selection tab could be added that allows you to go back to it without having to reselect the area 🤔

drogoganor commented:
I'll continue working on this but just letting you know I'm happy if you want to back out the editor selection PR until it's fully ready to go.

drogoganor commented:
I did the intended code refactor items I mentioned in a previous comment:
* Remove SelectedActorProperty from ActorEditLogic (source from Selection.Actor from EditorDefaultBrush instead)
* Set the EditorActorPreview.Selected property to true/false in the new EditorDefaultBrush.Selection setter instead.

I also applied your idea to invoke the SelectionChanged event from the EditorDefaultBrush.Selection setter instead, which is tidier and worked as intended.

Two annoyances still remain:
* Clicking a selection changed event in the History tab changes the tab to the (not shown) selection tab - as you mentioned, we could simply make the selection tab a visible tab instead. We have the room for it now. The tab could be disabled when there's nothing selected.
* When you delete a selected actor, it inserts two history events: unselect the actor, then delete it. It would be good if these were merged into one.

drogoganor commented:
![image](https://github.com/OpenRA/OpenRA/assets/1639681/07e97dcf-34be-4544-a2cd-482bcf3cef87)
Select tab button has been (re)introduced. It's disabled if there's no current selection.

Added fixes for the last two items mentioned:

* Clicking selection changed items in the history tab will no longer change to the Select tab. (Though this does produce an annoying issue - if you select an actor/area on the map with the history tab open, it won't go to the Select tab. Could be fixed with a parameter on the SelectionChanged event that determines whether the change was from undo/redo or from user action - however, the event is invoked from within the EditorDefaultBrush.Selection setter, so it will need some extra thought).
* Deleting a selected actor will now only produce one history item instead of two.

Also reintroduced a fix for issue #21314 as per PunkPun's suggested code change.

I also added a couple of little fixes for the new area selection info panel (I might look at that in more detail later).

I also made the chrome for the editor icon collection use a distinct set instead of reusing the timer and infantry icons as suggested in Discord by Dzierzan:
```
editor:
    Inherits: ^Glyphs
    Regions:
        tiles: 35, 171, 16, 16
        overlays: 52, 171, 16, 16
        tools: 69, 171, 16, 16
        actors: 100, 171, 16, 16
        history: 130, 171, 16, 16
```

Having a `ResourceLayer` defined should be optional again as per michael.dgg.2's suggestion.

ArtSrc changes are here: https://github.com/OpenRA/ArtSrc/pull/20

PunkPun commented on OpenRA.Mods.Common/EditorBrushes/EditorDefaultBrush.cs:
```diff
@@ -43,7 +45,32 @@ public sealed class EditorDefaultBrush : IEditorBrush
+				if (selectionInner.Actor != null)
+				{
+					selectionInner.Actor.Selected = true;
+				}
```
style nit: remove braces from both of these conditionals

PunkPun commented on OpenRA.Mods.Common/Widgets/Logic/Editor/MapEditorSelectionLogic.cs:
```diff
@@ -150,18 +148,20 @@ void HandleSelectionChanged()
+			DimensionsLabel.GetText = () => DimensionsAsString(selectionSize);
 			DiagonalLabel.GetText = () => $"{diagonalLength}";
-			ResourceCounterLabel.GetText = () => $"{resourceValueInRegion}$";
+			ResourceCounterLabel.GetText = () => $"{resourceValueInRegion:C}";
```
this is wrong, the ingame currency has the $ symbol and should not be localised

PunkPun reviewed (commented):
it looks like icons became off-centred

<img width="318" alt="Screenshot 2024-02-05 at 12 37 07" src="https://github.com/OpenRA/OpenRA/assets/37534529/57247e9a-5c01-46fe-b8b4-51fe21a40c4e">

when I select a resource, it blinks on the last spot where the add resource cursor was rendered. Same with tiles. I wonder if this is a recent regression.

I wonder how other cursors should work. Now when you change cursors you nolonger delete selection. The actor is still selected, and the area select still exists (you can copy paste), but just not rendered while another cursor is equipped. I feel like changing cursor should delete selection.

Perhaps the select tab should act as a button that sets cursor to the default cursor? That way it would still have a function while there's no selection available and would not need to be disabled

drogoganor commented:
Thanks for the code review again Pun. I agree, changing cursors should clear the selection - I added a `ClearSelection();` to the `EditorDefaultBrush.Dispose()` which does that.

I tested out the "flashing tile" issue on commit `b58c1ea` (before the selection changes) - it happens there, too. I guess it needs a tick to set the cursor position again.
Reproduction steps:
* Select a tile brush, and hover it somewhere
* Right-click to deselect the brush
* Select an actor brush
* Select a tile brush again

The select tab being enabled without a selection is a possibility - we could add an instructions label to the empty tab, to the effect of "Left-click to select an actor, or left-click and drag to select an area.". I guess the downside is that the tiles aren't shown as soon as you create a new map (likely the first thing the user will want to use).

Yes, it's a real pain that the icons seem off-center, too. I wonder if they can be offset in the chrome? The problem is that the buttons are 49 pixels wide.
EDIT: I'm an idiot - they have positions that are explicitly set! I will fix that.
EDIT: Button icons are centered again now.
![cnc-2024-02-06T090004692Z](https://github.com/OpenRA/OpenRA/assets/1639681/fdcd9c92-8bcb-46c8-97fa-8e24b753e7ac)

The area selection info panel needs a bit of work to fit correctly on the minimum supported resolution (1280x720 I believe), but I should probably leave that for another PR. I think the "Area Selection" header could have the dimensions and coordinate range instead, like "Area Selection: 13,47 - 25,55 (12x8)"
(Screenshot includes centered buttons)
![ra-2024-02-06T085932707Z](https://github.com/OpenRA/OpenRA/assets/1639681/9bca2b4b-2a19-43a2-8466-913d224fc838)

drogoganor commented:
I also made some adjustments to the area selection panel info area so it will fit at 1280x720. The area and dimensions are now shown in the title:

![ra-2024-02-06T093439269Z](https://github.com/OpenRA/OpenRA/assets/1639681/7157dbb5-e61c-4840-be11-ff3910fbae4b) ![cnc-2024-02-06T093414885Z](https://github.com/OpenRA/OpenRA/assets/1639681/67f4abb2-7592-42f7-91d0-63200fac5751)

I also correctly centered the title label.

drogoganor commented:
I've altered the area select behavior now so that the selection is kept when selecting a different brush. I know this is contrary to my earlier comment, but after playing around with it, I think it's better behavior.

Right clicking will not clear the selection. This matches actor selection behavior.
Right clicking an actor or overlay with an area selection will delete the actor or overlay.
Switching to the selection tab with another brush selected will no longer show a blank panel.

Undo/redo actions will no longer switch tabs. This required a change to IEditorAction.Do() to pass a `isUndoRedoAction` bool (Should this be named "isHistoryAction" instead?)

I'm much happier with how it all fits together now.

PunkPun commented on OpenRA.Mods.Common/EditorBrushes/EditorActorBrush.cs:
```diff
@@ -99,10 +99,10 @@ public AddActorAction(EditorActorLayer editorLayer, ActorReference actor)
 		}
 
-		public void Do()
+		public void Do(bool isUndoRedoAction)
```
`isUndoRedoAction` is overkill. Can't it be handled by Execute or something like that?

drogoganor replied:
Turns out a separate event triggered upon user actions was the way to go. This is also triggered by `ClearSelection` if the optional argument `updateSelectedTab` is set to true.

PunkPun reviewed (commented):
if you copy-paste with no filters on, the undo action won't undo the removed actors

drogoganor commented:
I reverted the change to the `IEditorAction` interface and introduced a separate event `EditorDefaultBrush.UpdateSelectedTab`, instead.
I also fixed the actor copy/paste not being gated by the filter correctly, ack.

PunkPun reviewed (approved):
LGTM

PunkPun commented:
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/61ede439d8f4ccb10288e683252392de95c3cd68...2797d5e16e2c0077add2438a426a1694ca39bfab)
